### PR TITLE
Rewrite `examples/Braccio_Learn_and_Repeat` using State design pattern

### PR DESCRIPTION
**Pull request by aentinger**
@Esquirio Can you please test?

### CODE REVIEW

**Comment by Esquirio** [comment]
In the first tests with this [code](https://github.com/arduino-libraries/Arduino_Braccio_plusplus/blob/af07b2ada0184915107ab8b46428009254c8c716/examples/Braccio_Learn_and_Repeat/Braccio_Learn_and_Repeat.ino), the firmware crashs in the REPLAY mode. Thus, it was necessary to restart the Arduino Board.

**Comment by clodpheasant** [comment]
@91volt @umbynos can you also test it out? 

I had no issues with the sketch actually

**Comment by aentinger** [comment]
Following up a private conversation with @Esquirio I think there is reason to believe that unprotected, multi-threaded access to LVGL (no `Braccio.lvgl_lock/unlock` statements) are cause for the spurious crashes.

@Esquirio can you please test the most recent code in this PR? Please ensure to download the whole library, not only the example (i.e. via this [link](https://github.com/arduino-libraries/Arduino_Braccio_plusplus/archive/refs/heads/main.zip)).

**Comment by Esquirio** [comment]
I've tested with the newest version of Braccio ++, and it presented the same problem. I am trying to check the code and find any issues

File: examples/Braccio_Learn_and_Repeat/Braccio_Learn_and_Repeat.ino [hunk 1]
```diff
@@ -25,10 +28,12 @@ static const char * btnm_map[] = { "RECORD", "\n", "REPLAY", "\n", "ZERO_POSITIO
 
 
 static void eventHandlerMenu(lv_event_t * e) {
+  Braccio.lvgl_lock();
   lv_event_code_t code = lv_event_get_code(e);
   lv_obj_t * obj = lv_event_get_target(e);
 
   if (code == LV_EVENT_KEY && lv_indev_get_key(lv_indev_get_act()) == LV_KEY_HOME) {
```
Comment (Esquirio) [inline examples/Braccio_Learn_and_Repeat/Braccio_Learn_and_Repeat.ino:35]: This if statement is crashing the code when I do as in this [PR](https://github.com/arduino-libraries/Arduino_Braccio_plusplus/blob/d8b4421c73bce5ea7fe3bc451592cd940928b4d8/examples/Braccio_Learn_and_Repeat/Braccio_Learn_and_Repeat.ino#L35), everything works fine, and the code didn't crash.

**Review by Esquirio — COMMENTED** [review]
The issue is in the if statement in lines 35-39. When I removed it, the example didn't crash.

**Comment by 91volt** [comment]
> Hi @Esquirio ☕ 👋
> 
> I've merged #72 and rebased this PR.
> 
> Can you give it a test and let me know if this works for you?
> 
> Btw, I've noticed that despite #72 claims to use ENTER for "clicking events" clicking the joystick still works just fine CC @91volt .

This was the intended behavior, both joystick click and enter button should allow the user to select an option.

**Comment by aentinger** [comment]
>This was the intended behavior, both joystick click and enter button should allow the user to select an option.
In this case all is fine :+1:

**Comment by 91volt** [comment]
@aentinger We have tested the PR, here me and @umbynos feedbacks:
- The `MAX_SAMPLES` lasts significantly more than 20 seconds.
- If record is stopped due to `MAX_SAMPLES` being reached, the REPLAY action doesn't work as expected and stops immediately.
- when max samples is reached, the ZERO_POSITION state is set the braccio doesn't actually go to zero, and remain disengaged. (Probably this happens because no lvgl events are triggered)

For what concerns the leds they work like a charm now.

**Comment by Esquirio** [comment]
I've tested the two codes intensively (@91volt code and @aentinger code). Both codes crashed. I will continue to investigate for issues and try to discover the cause of the crashes.

**Comment by aentinger** [comment]
The problem with this example is that

a) there are a fair number of state transitions
b) there are sometimes state enter/exit actions and that
c) state and stateful variables are both manipulated in loop and the LVGL event header.

I've tried avoiding it until now, but this whole examples needs a clean rewrite, otherwise it will be always be a trial-and-error programming with the smallest change breaking the whole flow.

**Comment by clodpheasant** [comment]
@91volt @Esquirio your 2 cents here?

**Comment by Esquirio** [comment]
@aentinger Did you mean the issue is in the state transitions in the display? 

I think it is possible to test the example using the Serial Monitor (or buttons and LEDs) instead of the LVGL library (If this test hasn't been done). I can do this test and report my findings.

**Comment by aentinger** [comment]
Once an application exceeds a certain number of states and state transitions it becomes very hard to keep track of the control flow is variables storing state are manipulated on several different locations in the code. Employing the [State](https://gameprogrammingpatterns.com/state.html) programming pattern I've untangled the transitions and make them more visible so that now there should be no more lurking :bug: .

@Esquirio :pray: test :bow:

**Pull request merged**